core/src/main/java/discord4j/core/shard/ShardingClientBuilder.java

```diff
@@ -0,0 +1,289 @@
+/*
+ * This file is part of Discord4J.
+ *
+ * Discord4J is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Discord4J is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Discord4J. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package discord4j.core.shard;
+
+import discord4j.common.JacksonResourceProvider;
+import discord4j.common.ReactorResourceProvider;
+import discord4j.core.DiscordClientBuilder;
+import discord4j.gateway.GatewayObserver;
+import discord4j.gateway.SimpleBucket;
+import discord4j.rest.RestClient;
+import discord4j.rest.http.ExchangeStrategies;
+import discord4j.rest.http.client.DiscordWebClient;
+import discord4j.rest.json.response.GatewayResponse;
+import discord4j.rest.request.DefaultRouterFactory;
+import discord4j.rest.request.Router;
+import discord4j.rest.request.RouterFactory;
+import discord4j.rest.request.SingleRouterFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.ReplayProcessor;
+import reactor.core.scheduler.Schedulers;
+import reactor.util.Logger;
+import reactor.util.Loggers;
+import reactor.util.function.Tuple2;
+
+import javax.annotation.Nullable;
+import java.time.Duration;
+import java.util.Objects;
+import java.util.function.Predicate;
+
+/**
+ * Builder suited for creating a sequence of {@link discord4j.core.DiscordClient} each dedicated to a shard, while
+ * allowing resource sharing and login coordination.
+ * <p>
+ * Creates multiple Discord clients with many configurable options. It uses a recommended amount of shards but can be
+ * overridden through {@link #setShardCount(Integer)}.
+ */
+public class ShardingClientBuilder {
```

Comment:
We should just remove getters on both this builder and DiscordClientBuilder. Getters on builders is an unusual pattern, at least I have not seen an example of it in a popular/enterprise project or even in the JDK itself (refer to HttpClient.Builder)

Plus I don't see much use cases considering you can't multi-thread these anyways.